{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import copy\n",
    "file = Path(\"hg38_coding_exons_bilan_ovlpTF_NR_cCREs_FANTOM5+Enh_NonCoding_TSS_TES_UTR5_UTR3_peakDensity_density50pb_AllSummitsPeaksDensitySum_correl_DnaseEncodeChipAtlas_ATAC.bed\")\n",
    "\n",
    "# We first take all tf and cells lines mapping on a gene \n",
    "dic_list = []\n",
    "with open(file) as f:\n",
    "    for line in f:      \n",
    "\n",
    "        if line.strip().split()[10] != \"NA\":\n",
    "\n",
    "            dic = {\"Exon ID\":\"\", \"Chrom\":\"\", \"Start\":\"\", \"End\":\"\", \"Nb_TF\":\"\",\"Density\":\"\", \"CDS Start\":False, \"CDS End\":False, \"UTR\":False, \"Number of TF < 10\":False, \"Number of TF >= 10\":False, \"Encode cCREs\":False, \"CTCF\":False, \"enhD\":False, \"enhP\":False, \"K4m3\":False, \"prom\":False, 'Fantom5 TSS':False, \"Fantom enhancer\":False, \"Non-coding first exon\":False, \"Exon with max density\":False, \"Max mapping density around exons (50bp)\":False, \"DNase ENCODE\":False,\"DNase ChipAtlas\":False,\"ATAC ChipAtlas\":False}\n",
    "\n",
    "            dic[\"Exon ID\"] = line.strip().split()[3]\n",
    "            dic[\"Chrom\"] = line.strip().split()[0]\n",
    "            dic[\"Start\"] = line.strip().split()[1]\n",
    "            dic[\"End\"] = line.strip().split()[2]\n",
    "            dic[\"Nb_TF\"] = line.strip().split()[10]\n",
    "\n",
    "            #Â check if first exon with cds start\n",
    "            cds_start = line.strip().split()[21]\n",
    "            if cds_start != \"NA\":\n",
    "                dic[\"CDS Start\"] = True\n",
    "\n",
    "            cds_end = line.strip().split()[22]\n",
    "            if cds_end != \"NA\":\n",
    "                dic[\"CDS End\"] = True\n",
    "            \n",
    "            UTR5 = line.strip().split()[23]\n",
    "            UTR3 = line.strip().split()[24]\n",
    "            if UTR5 != \"NA\" or UTR3 != \"NA\":\n",
    "                dic[\"UTR\"] = True\n",
    "\n",
    "            if line.strip().split()[33] != \"NA\":\n",
    "                dic[\"DNase ENCODE\"] = True\n",
    "\n",
    "            if line.strip().split()[35] != \"NA\":\n",
    "                dic[\"DNase ChipAtlas\"] = True\n",
    "\n",
    "            if line.strip().split()[40] != \"NA\":\n",
    "                dic[\"ATAC ChipAtlas\"] = True\n",
    "\n",
    "            # check other values\n",
    "            nb_tf = int(line.strip().split()[10])\n",
    "            if nb_tf < 10:\n",
    "                dic[\"Number of TF < 10\"] = True\n",
    "            else:\n",
    "                dic[\"Number of TF >= 10\"] = True\n",
    "            \n",
    "            ccre_label = line.strip().split()[16]\n",
    "            ccre_label_list = ccre_label.split(\",\")\n",
    "            for e in ccre_label_list:\n",
    "                if e != \"NA\":\n",
    "                    dic[\"Encode cCREs\"] = True\n",
    "                if e == \"CTCF\":\n",
    "                    dic[\"CTCF\"] = True\n",
    "                if e == \"enhD\":\n",
    "                    dic[\"enhD\"] = True\n",
    "                if e == \"enhP\":\n",
    "                    dic[\"enhP\"] = True\n",
    "                if e == \"K4m3\":\n",
    "                    dic[\"K4m3\"] = True\n",
    "                if e == \"prom\":\n",
    "                    dic[\"prom\"] = True\n",
    "\n",
    "            tss = line.strip().split()[17]\n",
    "            if tss != \"NA\":\n",
    "                dic[\"Fantom5 TSS\"] = True\n",
    "            \n",
    "            enhancer = line.strip().split()[19]\n",
    "            if enhancer != \"NA\":\n",
    "                dic[\"Fantom enhancer\"] = True\n",
    "            \n",
    "            #check if flag first exon of non coding gene\n",
    "            noncoding_gene = line.strip().split()[20]\n",
    "            if noncoding_gene != \"NA\":\n",
    "                noncoding_gene_list = noncoding_gene.split(\",\")\n",
    "                for e in noncoding_gene_list:\n",
    "                        dic[\"Non-coding first exon\"] = True\n",
    "\n",
    "            #check if exon has the most density of peak chipseq all in it\n",
    "            peak_exon = line.strip().split()[25]\n",
    "            dic[\"Density\"] = peak_exon\n",
    "            peak_50pb = line.strip().split()[26]\n",
    "            if int(peak_exon) >= int(peak_50pb):\n",
    "                dic[\"Exon with peak density\"] = True\n",
    "            else:\n",
    "                dic[\"Max mapping density around exons (50bp)\"] = True\n",
    "            \n",
    "            \n",
    "            #save row\n",
    "            dic_list.append(copy.deepcopy(dic))\n",
    "\n",
    "\n",
    "df = pd.DataFrame(dic_list)\n",
    "data = [\n",
    "    {\"labels\": [\"Promoter\",\"Enhancer proximal\",\"Enhancer distal\",\"H3K4me2\",\"DNAse-seq\",\"H3K4me1\",\"H3K27ac\"], \"values\": [df['prom'].value_counts().loc[True],df['enhP'].value_counts().loc[True],df['enhD'].value_counts().loc[True],42747,df['DNase ENCODE'].value_counts().loc[True],103314,134460], \"colors\": [\"red\",\"orange\",\"yellow\",\"cyan\",\"limegreen\",\"deepskyblue\",\"blue\"]},                # First subplot with 1 bar\n",
    "    {\"labels\": [\"H3K27ac\"], \"values\": [54375], \"colors\": [\"blue\"]}, # Second subplot with 3 bars\n",
    "    {\"labels\": [\"DNAse-seq\",\"ATAC-seq\"], \"values\": [df['DNase ChipAtlas'].value_counts().loc[True],df['ATAC ChipAtlas'].value_counts().loc[True]], \"colors\": [\"limegreen\",\"green\"]},\n",
    "    {\"labels\": [\"CAGE TSS\"], \"values\": [df['Fantom5 TSS'].value_counts().loc[True]], \"colors\": [\"darkorchid\"]},\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import matplotlib.ticker as ticker\n",
    "import seaborn as sns\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "from matplotlib.gridspec import GridSpec\n",
    "\n",
    "titles = [\"ENCODE\",\"GTEx\",\"CHIP-ATLAS\",\"FANTOM\"]\n",
    "\n",
    "# Calculate the relative heights for each subplot\n",
    "heights = [len(d[\"labels\"]) for d in data]  # Heights proportional to the number of bars\n",
    "\n",
    "# Create a gridspec with proportional subplot heights\n",
    "fig = plt.figure(figsize=(5, 5))\n",
    "gs = GridSpec(len(data), 1, height_ratios=heights)\n",
    "\n",
    "# Plot each dataset\n",
    "for i, d in enumerate(data):\n",
    "    ax = fig.add_subplot(gs[i])\n",
    "    bars = ax.barh(d[\"labels\"], d[\"values\"], color=d[\"colors\"])  # Horizontal bar plot\n",
    "    ax.set_title(f\"{titles[i]}\")\n",
    "    \n",
    "    # Ensure consistent spacing by adjusting y-limits\n",
    "    n_bars = len(d[\"labels\"])\n",
    "    ax.set_ylim(-0.5, n_bars - 0.5)  # Add uniform padding around the bars\n",
    "\n",
    "    ax.set_xlim(0, 165000)  # Align x-axis limits\n",
    "    ax.invert_yaxis()\n",
    "    ax.set_xticks([])\n",
    "\n",
    "    # Add value labels in front of each bar\n",
    "    for bar, value in zip(bars, d[\"values\"]):\n",
    "        ax.text(value, bar.get_y() + bar.get_height() / 2,  # Position slightly ahead of the bar\n",
    "                f\" {value}\", va='center', ha='left')  # Align text to the left of the bar\n",
    "\n",
    "# Common x-label\n",
    "ax.set_xticks([0,40000,80000,120000,160000])\n",
    "plt.xlabel(\"Count\")\n",
    "plt.tight_layout()\n",
    "\n",
    "fig.suptitle(\"Count of hg38 merged exons overlapping ressources\", y=1.05)\n",
    "plt.savefig(\"/home/mouren/Images/tls1/overlap_exons_with_all_catalogs/all_exons_ovlp_bilan_fig1_2.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
